USWDS-Site - Card: Fix card grid presentation

amyleadem opened the pull request:
# Summary
- **Switched out the grid classes in the card preview template.** This improves the appearance of the card component preview at narrow widths.
- **Added grid styles specifically suited for the card component preview.**
- **Added a changelog entry for card grid fix.**

## Related issue

Related to https://github.com/uswds/uswds-site/issues/2323

## Preview link

- [Card component preview](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/al-card-grid-fix/components/card/#card-preview-content)
- [Card changelog ](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/al-card-grid-fix/components/card/#latest-updates)

## Problem statement

The [card component preview page](https://designsystem.digital.gov/components/card/) shows cards that get too narrow and elements within the cards overlap at narrow browser widths. 

![image](https://github.com/uswds/uswds/assets/93996430/829f2f51-e5ee-410c-94cf-83bd3863cdde)

## Solution
Adjusting the grid breakpoints in the card template improves the display of the component at narrow widths. 

This fix currently provides specific grid classes for only for the scenarios needed. Alternatively, we could turn on the "widescreen" and "tablet-lg" values in `$theme-utility-breakpoints`. However, this felt heavy-handed for this specific need (See more info in [this comment](https://github.com/uswds/uswds-site/pull/2299#issuecomment-1819706789)). 

## Testing and review
- Confirm the [card component preview](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/al-card-grid-fix/components/card/#card-preview-content) presents well at various breakpoints and has the correct grid breakpoints. 
- Confirm that the approach for updating the styles makes sense. 
- Confirm the changelog is accurate and free from spelling and grammatical errors.

mahoneycm reviewed (changes requested):
Ran into an issue trying to get my local to build. Also just left a question about the approach and the alternative you mentioned in the description.

mejiaj commented on css/_uswds-theme-custom-styles.scss:
```diff
@@ -37,6 +37,26 @@ $site-measure: 5;
   @include usa-grid-col;
 }
 
+// Specific grid-col styes added for use in the card component grid
+@media all and (min-width: 55em){
+  .tablet-lg\:grid-col-6 {
+    flex: 0 1 auto!important;
+    width: 50%!important;
+  }
+}
+
+@media all and (min-width: 87.5em){
+  .widescreen\:grid-col-4 {
+    flex: 0 1 auto!important;
+    width:33.3333333333%!important;
+  }
+
+  .widescreen\:grid-col-6 {
+    flex: 0 1 auto!important;
+    width: 50%!important;
+  }
+}
+
```
Might be more straightforward to follow a similar pattern to Banner and use the replace filter to change the class.

https://github.com/uswds/uswds-site/blob/1fec7d4889b34d0be1ed55fd9f5c10b157b06be9/_includes/code/components/banner.html#L2-L8

**Example `/_includes/code/components/card.html`**

```liquid
{% capture card--flag %}
  {% library_component usa-card~flag %}
{% endcapture %}{{ card--flag
  | strip
  | replace: 'desktop:grid-col-6', 'widescreen:grid-col-6'
}}
```

**Gives us expected result:**
<img width="900" alt="image" src="https://github.com/uswds/uswds-site/assets/3385219/fc23e6ff-a601-4153-bd0d-b896b0f5ae6e">

mahoneycm replied:
Woah that's an interesting alternative! Helps the solution to stay site specific which is nice.

amyleadem replied:
@mejiaj Thanks for that suggestion! Much cleaner. I updated the templates in [c6dd23d](https://github.com/uswds/uswds-site/pull/2299/commits/c6dd23da7a445ad3ff4aec32e577b05a4451153d)

mejiaj replied:
Do we need the styles since its in the example container and that's a fixed width?

amyleadem replied:
@mahoneycm noted that if we don't want to change the Sass to bring in `widescreen` and `tablet-lg` breakpoints, we can use this code instead:
```html
<h3 class="site-preview-heading">Default card</h3>
{% capture card--standard %}
  {% library_component usa-card~standard %}
{% endcapture %}{{ card--standard
  | strip
  | replace: 'tablet:grid-col-4', 'tablet:grid-col-6 desktop-lg:grid-col-4'
}}

<h3 class="site-preview-heading">Flag layout card</h3>
{% capture card--flag %}
  {% library_component usa-card~flag %}
{% endcapture %}{{ card--flag
  | strip
  | replace: 'desktop:grid-col-6', 'flex-1'
}}
```

The breakpoints aren't quite as clean visually, but it is a noticeable improvement. This code prevents the need to either duplicate USWDS style rules in our project Sass OR add a bunch of extra CSS by turning on the two breakpoints via settings.

My personal preference is still to bring in the three grid style rules to our Sass. I know it is not ideal for code standards but for the best visual presentation I want to optimize the breakpoints as much as possible. I'd also like to see a possible feature request to be able to turn on breakpoints specifically for grid (or possibly turn it on for each individual utility), so this could be considered a stopgap measure if we decide to go down that path in the future. @mejiaj and @mahoneycm curious what you think!

amyleadem replied:
I noted my decisions on this topic in [this comment](https://github.com/uswds/uswds-site/pull/2299#issuecomment-1819706789) for visibility.

mejiaj reviewed (changes requested):
Added a possible alternative to adding more to custom SCSS.

amyleadem commented on package.json:
```diff
@@ -101,6 +101,6 @@
   "dependencies": {
     "@uswds/compile": "^1.0.0",
-    "@uswds/uswds": "^3.6.1"
+    "@uswds/uswds": "github:uswds/uswds#al-card-grid-fix"
```
Reverted in [571e7b2](https://github.com/uswds/uswds-site/pull/2299/commits/571e7b211724c8af5326d66389ce9d16d43d3428).

amyleadem commented:
@mejiaj @mahoneycm 
I have updated the grid classes in this repo and removed the USWDS test branch. I think we can now safely remove this item from the 3.7.0 release!

I added a note about why I didn't turn on the breakpoint via settings in [this comment](https://github.com/uswds/uswds-site/pull/2299#discussion_r1372363968). Let me know if you see a better way to do this.

amyleadem commented on css/_uswds-theme-custom-styles.scss:
I ran prettier to fix most of these but still get the BEM errors. We can figure those out once we finalize the approach for turning on the breakpoints.

amyleadem replied:
Noting that we can disable stylelint checks for these particular rules to prevent errors (if we decide to go this direction).

amyleadem replied:
Added stylelint ignore comments in https://github.com/uswds/uswds-site/pull/2299/commits/21e3a568f10f3769d8600baff6edecb6b548d051

mahoneycm reviewed (changes requested):
Hey Amy!

I found a way to achieve the same styles using the string replacement without the need for any scss changes. This route resolves the build errors and isolates the changes.

Let me know what you think about this approach!

amyleadem commented on _includes/code/components/card.html:
```diff
@@ -1,9 +1,15 @@
+{% endcapture %}{{ card--flag
+  | strip
+  | replace: 'desktop:grid-col-6', 'widescreen:grid-col-6'
+}}
```
Updated the markup replacement in https://github.com/uswds/uswds-site/pull/2299/commits/f012b2b9a1752aed0209050f263a78628597a97f

amyleadem commented on css/_uswds-theme-custom-styles.scss:
```diff
@@ -37,6 +37,26 @@ $site-measure: 5;
+    width: 50% !important;
+  }
+}
+
```
Removed `.widescreen\:grid-col-6` rule and added stylelint ignore comments in [21e3a56](https://github.com/uswds/uswds-site/pull/2299/commits/21e3a568f10f3769d8600baff6edecb6b548d051)

amyleadem commented on _includes/code/components/card.html:
```diff
@@ -1,9 +1,15 @@
 {% capture card--standard %}
   {% library_component usa-card~standard %}
-{% endcapture %}{{ card--standard | strip }}
+{% endcapture %}{{ card--standard
+  | strip
+  | replace: 'tablet:grid-col-4', 'tablet-lg:grid-col-6 widescreen:grid-col-4'
+}}
```
We determined that we still needed grid class replacement for the standard cards.

amyleadem commented:
The comment history here has gotten a bit jumbled, but I wanted to summarize the recent changes in this PR:
- I kept the `tablet-lg:grid-col-6 widescreen:grid-col-4` classes on `card--standard`. These grid breakpoints are not included in our breakpoint list, which means that we need copy over a couple of grid style rules from USWDS. However, I felt that these were the ideal breakpoints for the standard cards, and since it is a commonly-used component, I wanted to showcase them in their best light. 
    - Alternatively, we could use `tablet:grid-col-6 desktop-lg:grid-col-4`, which are already accounted for in the CSS. No additional CSS rules would be required if we chose these classes instead, but the cards do end up getting a bit narrow at various widths. 
    - Alternatively, we could also pursue custom site-specific class names here instead of copying over USWDS rules, but I went with the standard USWDS grid classes so that users would be able to use the code if they copied it from web inspector. 
    - We could also turn these breakpoints in the `$theme-utility-breakpoints` setting, but this would add a couple thousand more references and 100kb+ to the compiled CSS. 
- These grid classes do not meet BEM standards, so I had to exclude them from stylelint checks by using `stylelint-disable` comments. 
- I changed the `card--flag` classes to `flex-1` as @mahoneycm suggested. This allowed for the desired presentation without needing to add an additional rule definition for `.widescreen\:grid-col-6`.

amyleadem commented:
@mejiaj  @mahoneycm ready for your re-review. Let me know if you have comments.

mahoneycm reviewed (approved):
Lgtm!

As long as we're ok with maintaining the CSS overrides, I think that this is a visual improvement to showcase cards 👍 

- [x] Cards break at logical breakpoints
- [x] Cards do not feel two thing at any point
- [x] No errors when running `npm run build:css`

mejiaj reviewed (approved):
LGTM, glad we were able to agree on a fix. Good to merge once builds pass.
